src/peakrdl_sv/reg_top.sv.tpl

```diff
@@ -203,10 +203,9 @@ module ${lblock}_reg_top
     addr_hit = '0;
     % for i,r in enumerate(registers):
 <%
-    write_enables = reg_enables[r.path.lower()]['we']
     base_idx = reg_enables[r.path.lower()]['idx']
 %>\
-    % for i,enable in enumerate(write_enables):
+    % for i in range(len(r)):
```

Comment:
This won't work either as it doesn't take into account the sub-registers within wide regs.  You want to use the `num_regs` term that is calculated in the header.  I think we only need this value in `reg_top.sv.tpl` but if we also need it in the package template, then we should migrate the logic into the Python classes.

Reply:
But given we need to iterate over registers, you probably have to redo the iterator yourself or add an `addr_hit` key to the `reg_enables` dict and do it that way.

Reply:
> This won't work either as it doesn't take into account the sub-registers within wide regs

I believe it does. If a register is wide it will contain multiple fields that span beyond the access width. By iterating over the `len(r)` we get a relative index `i` to each field, which resolves to an asbolute index when you do `justified = "{}".format(base_idx+i).rjust(max_regs_char)` 

For the case of Phaser, with this change the generated output is 

```
logic [43:0] addr_hit;
always_comb begin
  addr_hit = '0;
  addr_hit[ 0] = (reg_addr == PHASER_MVP_REGISTERS_VERSION_BOARD_ID_OFFSET);
  addr_hit[ 1] = (reg_addr == PHASER_MVP_REGISTERS_VERSION_HW_REV_OFFSET);
  addr_hit[ 2] = (reg_addr == PHASER_MVP_REGISTERS_VERSION_GW_REV_OFFSET);
  addr_hit[ 3] = (reg_addr == PHASER_MVP_REGISTERS_CONFIG_CLK_SEL_OFFSET);
  addr_hit[ 4] = (reg_addr == PHASER_MVP_REGISTERS_CONFIG_LED_OFFSET);
  addr_hit[ 5] = (reg_addr == PHASER_MVP_REGISTERS_CONFIG_FAN_OFFSET);
  addr_hit[ 6] = (reg_addr == PHASER_MVP_REGISTERS_FASTLINK_CRC_ERR_OFFSET);
  addr_hit[ 7] = (reg_addr == PHASER_MVP_REGISTERS_SPI_IDLE_OFFSET);
  addr_hit[ 8] = (reg_addr == PHASER_MVP_REGISTERS_SPI_CFG_OFFSET);
  addr_hit[ 9] = (reg_addr == PHASER_MVP_REGISTERS_SPI_CFG_OFFSET);
  addr_hit[10] = (reg_addr == PHASER_MVP_REGISTERS_SPI_CFG_OFFSET);
  addr_hit[11] = (reg_addr == PHASER_MVP_REGISTERS_SPI_CFG_OFFSET);
  addr_hit[12] = (reg_addr == PHASER_MVP_REGISTERS_SPI_CFG_OFFSET);
  addr_hit[13] = (reg_addr == PHASER_MVP_REGISTERS_SPI_CFG_OFFSET);
  addr_hit[ 9] = (reg_addr == PHASER_MVP_REGISTERS_SPI_DIVLEN_OFFSET);
  addr_hit[10] = (reg_addr == PHASER_MVP_REGISTERS_SPI_SEL_OFFSET);
  addr_hit[11] = (reg_addr == PHASER_MVP_REGISTERS_SPI_DATW_OFFSET);
  addr_hit[12] = (reg_addr == PHASER_MVP_REGISTERS_SPI_DATR_OFFSET);
  addr_hit[13] = (reg_addr == PHASER_MVP_REGISTERS_ATTENUATOR_0_CFG_OFFSET);
  addr_hit[14] = (reg_addr == PHASER_MVP_REGISTERS_ATTENUATOR_1_CFG_OFFSET);
  addr_hit[15] = (reg_addr == PHASER_MVP_REGISTERS_ADC_CFG_OFFSET);
  addr_hit[16] = (reg_addr == PHASER_MVP_REGISTERS_ADC_CFG_OFFSET);
  addr_hit[16] = (reg_addr == PHASER_MVP_REGISTERS_ADC_STATUS_OFFSET);
  addr_hit[17] = (reg_addr == PHASER_MVP_REGISTERS_ADC_STATUS_OFFSET);
  addr_hit[17] = (reg_addr == PHASER_MVP_REGISTERS_DAC_ALARM_OFFSET);
  addr_hit[18] = (reg_addr == PHASER_MVP_REGISTERS_DAC_CFG_OFFSET);
  addr_hit[19] = (reg_addr == PHASER_MVP_REGISTERS_DAC_CFG_OFFSET);
  addr_hit[20] = (reg_addr == PHASER_MVP_REGISTERS_DAC_CFG_OFFSET);
  addr_hit[19] = (reg_addr == PHASER_MVP_REGISTERS_DAC_WRITEBACK_DATA_0_OFFSET);
  addr_hit[20] = (reg_addr == PHASER_MVP_REGISTERS_DAC_WRITEBACK_DATA_1_OFFSET);
  addr_hit[21] = (reg_addr == PHASER_MVP_REGISTERS_DAC_WRITEBACK_DATA_2_OFFSET);
  addr_hit[22] = (reg_addr == PHASER_MVP_REGISTERS_DAC_WRITEBACK_DATA_3_OFFSET);
  addr_hit[23] = (reg_addr == PHASER_MVP_REGISTERS_DAC_TEST_DATA_0_OFFSET);
  addr_hit[24] = (reg_addr == PHASER_MVP_REGISTERS_DAC_TEST_DATA_1_OFFSET);
  addr_hit[25] = (reg_addr == PHASER_MVP_REGISTERS_DAC_TEST_DATA_2_OFFSET);
  addr_hit[26] = (reg_addr == PHASER_MVP_REGISTERS_DAC_TEST_DATA_3_OFFSET);
  addr_hit[27] = (reg_addr == PHASER_MVP_REGISTERS_DAC_SYNC_DLY_OFFSET);
  addr_hit[28] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_CFG_OFFSET);
  addr_hit[29] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_CFG_OFFSET);
  addr_hit[30] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_CFG_OFFSET);
  addr_hit[29] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_FREQUENCY_0_OFFSET);
  addr_hit[30] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_FREQUENCY_1_OFFSET);
  addr_hit[31] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_FREQUENCY_2_OFFSET);
  addr_hit[32] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_FREQUENCY_3_OFFSET);
  addr_hit[33] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_PHASE_0_OFFSET);
  addr_hit[34] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_PHASE_1_OFFSET);
  addr_hit[35] = (reg_addr == PHASER_MVP_REGISTERS_DUC_0_STB_OFFSET);
  addr_hit[36] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_CFG_OFFSET);
  addr_hit[37] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_CFG_OFFSET);
  addr_hit[38] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_CFG_OFFSET);
  addr_hit[37] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_FREQUENCY_0_OFFSET);
  addr_hit[38] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_FREQUENCY_1_OFFSET);
  addr_hit[39] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_FREQUENCY_2_OFFSET);
  addr_hit[40] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_FREQUENCY_3_OFFSET);
  addr_hit[41] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_PHASE_0_OFFSET);
  addr_hit[42] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_PHASE_1_OFFSET);
  addr_hit[43] = (reg_addr == PHASER_MVP_REGISTERS_DUC_1_STB_OFFSET);
end
```

We see here the `duc_0.frequency` is wide 

```
reg {
    desc = "DUC frequency tuning word";
    regwidth = 32;

    field {} data0[7:0];
    field {} data1[15:8];
    field {} data2[23:16];
    field {} data3[31:24];
} frequency;
```

and the above output has an individual `addr_hit` entry for each of the four fields. I believe this is what you mean but if I have interpreted it wrong let me know.

Reply:
It works in that case because you have a single field per sub-reg.  If you had an accesswidth of 8 and 8 single bit fields, you'd end up with `len(r)` returning 8 and generating 8 addr_hit terms for a single access width.  You need to iterate over `r.subregs` I think.

Reply:
Ah I see the issue now, i'll try to provide a fix.